### ISSUE TITLE
Show window controls when idle

### ISSUE THREAD

**Issue by Keith94** [Samillion/ModernZ#251]
### Expected behavior of the wanted feature

Enable `window_top_bar` functionality when mpv is on idle screen (no file loaded).

Workaround for enabling **native** title bar only on idle screen: https://github.com/Samillion/ModernZ/issues/251#issuecomment-4019194253

**Comment by Samillion** [comment]
Oh wow, I didn't even realize this doesn't show, even on stock osc! I wonder why.

**Comment by Keith94** [comment]
> Oh wow, I didn't even realize this doesn't show, even on stock osc! I wonder why.

Yeah it's strange indeed; there's a feature request for it here: https://github.com/mpv-player/mpv/issues/13167

**Comment by Anthonyy232** [comment]
This would actually be awesome.

**Comment by zenwav** [comment]
I second this.

**Comment by Samillion** [comment]
I'm still trying to figure out why mpv hasn't implemented this by default in `osc.lua`. I mentioned it a couple of times on irc, but I wasn't always present the last month.

**Comment by Keith94** [comment]
True. Would be simple to port it over once they add it officially.

**Comment by Samillion** [comment]
Indeed. One of the main things I wanted to keep with ModernZ is that it's based on `osc.lua`.

This specific feature would require a drastic change (osc render). Unless there is a simple way that my brain is unable to think of yet.

**Comment by Keith94** [comment]
There's another popular 3rd party OSC that implemented it, but I didn't look into how he did it.

**Comment by Keith94** [comment]
I found some property called idle-active but no idea if that's even helpful.

`mp.observe_property('idle-active', 'bool', function(_, idle)`

[source](https://github.com/tomasklaen/uosc/commit/b8a37ef576cf51b33ab859012245cc787cefe536)

**Comment by Samillion** [comment]
> There's another popular 3rd party OSC that implemented it, but I didn't look into how he did it.

What's their repo link? I'd love to see their implementation.

**Comment by Keith94** [comment]
https://github.com/tomasklaen/uosc

**Comment by Samillion** [comment]
Ohh. Yeah. To be very frank, uosc is an amazing implementation as an osc overall that broke the osc standard and went over and beyond. It is the closest thing to a GUI toolkit yet it's just an OSC.

To be clear, this feature is doable. A simple (broken) example, in the tick function:
https://github.com/Samillion/ModernZ/blob/80ea6b715412ea339ad8097a4e8c7cc014d344a7/modernz.lua#L3112-L3117

It controls osc render:
https://github.com/Samillion/ModernZ/blob/80ea6b715412ea339ad8097a4e8c7cc014d344a7/modernz.lua#L3162-L3170

Changing this only shows it (broken) with osc in idle:
```diff
        end
+    end
-    elseif (state.fullscreen and user_opts.showfullscreen)
+    if (state.fullscreen and user_opts.showfullscreen)
```

Obviously it's broken here because it's essentially ticking render constantly, but it shows you why the osc would need many changes.

- Adjust `show/hide_osc` behavior for idle state and include osc `render/render_wipe`
- Adjust default values (such as title, play controls and such) to match an idle state
- Adjust `show/hide_osc` to be separate for top bar and controls (bottom)

The end result though would be many changes that would break the "based on osc.lua" thing. I think this is me hesitating because of paranoia and false set of standards.

Also, I'm almost certain there is even a simpler way to do this that my brain is just not picking up lol.

**Comment by Keith94** [comment]
Damn. That could explain why it hasn't made it into mpv yet because of all the changes. Probably not a lot of discussion around it either.

Should we close and re open it later once it's added?

**Comment by Samillion** [comment]
I don't mind leaving it open, since the same mpv issue has been open for over a year now. Maybe this will allow it to be discovered more and someone shares an input or an idea that make the implementation of this feature not so complicated..

**Comment by Samillion** [comment]
The context menu has made this possibly easier to implement. Since all that's really needed for ease of use is window controls, right?

So in `tick = function()` inside `if state.idle then`, we can add something like:
```lua
        -- window controls
        local btn_y = 8
        local btn_spacing = 38
        local right = display_w - 15

        local function wc_button(x, glyph)
            ass:new_event()
            ass:pos(x, btn_y)
            ass:an(9)
            ass:append(string.format("{\\fn%s\\fs20\\bord0}%s", iconfont, glyph))
        end

        if not icons then
            set_icon_theme()
        end

        -- close
        wc_button(right, icons.window.close)

        -- maximize / restore
        if state.fullscreen then
            wc_button(right - btn_spacing, icons.window.unmaximize)
        else
            wc_button(right - btn_spacing, icons.window.maximize)
        end

        -- minimize
        wc_button(right - btn_spacing * 2, icons.window.minimize)
```

But then we'd have to add keybinds and areas to make them actually clickable.

<img width="1717" height="967" alt="Image" src="https://github.com/user-attachments/assets/990745fb-b607-4197-8171-f3be7d5a1401" />

**Comment by Keith94** [comment]
>Since all that's really needed for ease of use is window controls, right?

Good point. That would be best.

Even if you were to show an idle window title saying just "mpv" that wouldn't add much to the user experience.

**Comment by Samillion** [comment]
I can get it functioning, but honestly, I hate how I'm implementing it. My code for this seem so bulky. I'll crawl in another cave for a while to think of a better way.

Just to demonstrate though, it is doable. It just needs a better brain than mine lol.

https://github.com/user-attachments/assets/e2ce1ffc-e7d6-4dab-be09-1784fd87ce00

**Comment by Keith94** [comment]
Wow, that looks awesome; any implementation is better than none 😛

Too bad it can't be a simple fix. "If state idle then enable window controls" or something.

**Comment by Samillion** [comment]
> Too bad it can't be a simple fix. 

Indeed. The problem is, my implementation seems practical, but it's not even done and it's like 80+ lines. The area and keybind parts alone look like a nightmare.

If you notice, the window controls in idle are not scaled in any way like the actual osc. So adding scale to it would mean even more bulk.

Keep in mind, I'm not saying it has to be bloated. I'm saying my brain can only come up with a bloated way... so far lol.

> "If state idle then enable window controls" or something.

Probably possible. Someone use AI, let it do that thing and save me the little brain power I have lol.

**Comment by Keith94** [comment]
It's still an accomplishment to have gotten it working, even before mpv itself.
Also, is that a Windows skin? Powershell window looks so sleek.

**Comment by Samillion** [comment]
> It's still an accomplishment to have gotten it working, even before mpv itself. 

❤️ 

> Also, is that a Windows skin? Powershell window looks so sleek.

Just native [Terminal](https://apps.microsoft.com/detail/9n0dx20hk701) app, with `One Half Dark` color scheme.

<img width="1483" height="762" alt="Image" src="https://github.com/user-attachments/assets/7c33f96b-c912-4cbc-ab95-bee32a652a4e" />

**Comment by Keith94** [comment]
That makes sense. I thought it was a mod/skin because my window opacity doesn't look like that lol.

**Comment by Keith94** [comment]
Um, does this help at all? It's an obvious AI response..

---

## The core problem

When `state.idle == true`, `tick()` takes the **idle branch** — it renders the logo and then **explicitly disables `showhide` key bindings** (lines 3802–3806). It never calls `render()`, which is the only place window controls are drawn, faded, and their mouse input areas activated. So the window controls are completely dead in idle mode.

---

## How to enable window controls in `state.idle`

You need to make two targeted changes inside the idle branch of `tick()`:

### Change 1 — Don't disable the `showhide_wc` binding in idle

```lua
-- CURRENT (lines 3802–3806): kills both bindings
if state.showhide_enabled then
    mp.disable_key_bindings("showhide")
    mp.disable_key_bindings("showhide_wc")   -- ← remove this line
    state.showhide_enabled = false
end
```

### Change 2 — Call a minimal window-controls render after the idle OSD

Right after `set_osd(...)` on line 3800, add a call to render just the window controls:

```lua
set_osd(display_w, display_h, ass.text, -1000)

-- ADD THIS: render window controls even during idle
if window_controls_enabled() then
    local wc_vis
    if user_opts.independent_zones then
        wc_vis = state.wc_visible
    else
        wc_vis = true
    end
    if wc_vis then
        render_elements(ass, false, true)   -- osc_vis=false, wc_vis=true
    end
end
```

> **Note:** `render_elements(ass, false, true)` is already designed to skip bottom OSC elements when `osc_vis` is `false` — it only draws elements whose `layout.group == "top"`. So this is safe and won't draw the seekbar or buttons over the idle logo.

---

## Why `window_top_bar` setting matters

The `window_controls_enabled()` function (line 989) controls whether the top bar renders at all:

| `window_top_bar` setting | Idle window controls will show? |
|---|---|
| `"auto"` | ✅ Yes — when borderless or fullscreen |
| `"yes"` | ✅ Always |
| `"no"` | ❌ Never |

If you want them unconditionally in idle, you can bypass `window_controls_enabled()` and just always call `render_elements(ass, false, true)` in the idle branch.

---

## Summary of the call chain

```
tick()
 └─ state.idle == true
     ├─ renders logo ASS          ← happens now
     ├─ disables showhide_wc      ← REMOVE this
     └─ never calls render()      ← ADD wc render here
         └─ render_elements(ass, false, true)
             └─ draws group=="top" elements only (window controls)
```

**Comment by Samillion** [comment]
The text makes sense, the implementation doesn't.

The way I did it is similar

```lua
-- in tick
render_idle_window_controls(ass, display_w, display_h)
```

But, the function includes a whole new draw event for the controls. Which I then verified by adding a new state for window controls in idle

And that's just drawing the controls. Next you need to assign keybinds so the buttons actually function like they are on ModernZ, and that I'm not even going to show how I did it lol, it's just terrible.

THEN you have to add scaling to those controls, otherwise they'll always be huge.

THEN you have to delete all my code and redo it in like 5 lines lol

**Comment by Samillion** [comment]
@Keith94 I'm curious, which AI agent was that? I was honestly impressed by the work in independent zones original PR, but this one seems to lack actual understanding of the code in ModernZ by the AI agent.

**Comment by Keith94** [comment]
That's def a lot of code you posted, but could be nice for temporary solution.

I used claude.ai, the free version. It's based on just a single prompt, though, so there's not much context for it to go off, I guess.

**Comment by Keith94** [comment]
Off topic. I thought of a way to get the native (not OSC) title bar working; just use conditional profile.

```editorconfig
[Idle]
    profile-desc=Options when mpv is in idle state
    profile-cond=idle and p["idle-active"]
    profile-restore=copy-equal
    title=mpv
    title-bar=yes
```
(thanks to Sam for updated profile)

**Comment by Samillion** [comment]
> Off topic. I thought of a way to get the native (not OSC) title bar working; just use conditional profile.

Isn't this.... doesn't this solve everything? lol That is NOT offtopic.

The easiest most obvious solution, yet it escaped me all this time. Well done. Literally does what's requested here.

Adding it to my [`mpv.conf`](https://github.com/Samillion/mpv-conf/blob/master/mpv.conf) right now.

```EditorConfig
[Idle]
    profile-desc=Options when mpv is in idle state
    profile-cond=idle and p["idle-active"]
    profile-restore=copy-equal
    title=mpv
    title-bar=yes
```

**Comment by Keith94** [comment]
Do we consider it solved? I think you can close if you want.

**Comment by Samillion** [comment]
I have no idea to be honest. It definitely solves the issue. But this tiny piece of crap geek inside of me is like "it's not the same" lol.

I don't mind leaving it open. More minds to see and think of an efficient solution.

**Comment by Samillion** [comment]
Idle Profile: Seems `or` was causing issues. Replaced with `and`.

**Comment by Keith94** [comment]
> Idle Profile: Seems `or` was causing issues. Replaced with `and`.

This one works perfectly.

On mpv side, it doesn't know how to keep the window centered on file load if it goes from `title-bar=yes` (idle) to `title-bar=no` (file loaded). But that's really minor side effect.

**Comment by Samillion** [comment]
https://github.com/mpv-player/mpv/pull/17677 🥳 

See what I mean? My implementation was so far off, I could not think of it in this simple manner. Mine added specific draw helpers, input helpers...etc just for idle state 😆 

Will apply upstream feat soon.

**Comment by Keith94** [comment]
Oh wow, that is awesome. All just a matter of time lol. Thanks for being on top of it.
